ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -117,11 +117,56 @@ As a third-party application developer that uses shipwright under the hood, I wa
+
+```go
+type FailureDetails struct {
+    Reason   string    `json:"reason,omitempty"`
+    Message  string    `json:"message,omitempty"`
+    Location *FailedAt `json:"location,omitempty"`
+}
+``` 
+
+Second, we keep the `failedAt` field to prevent breaking user applications. However, it should be deprecated and removed in the future due to its redundancy. Therefore, we propose adding documents that state the deprecation status and adding the news to the release notes.
+To ease deprecation and future removal, we propose refactoring `UpdateBuildRunUsingTaskRunCondition` by extracting the logic of finding the failed pod and container in a BuildRun/TaskRun.
+The new file `pkg/reconciler/buildrun/resources/failures.go` includes the extraced logic and other failure related functions.
+
+A TaskRun can emit failures using two results:
+1. shp-error-reason
+2. shp-error-message
+
+Whenever a Tekton TaskRun fails, it keeps a JSON string of all results inside each terminated step in `TaskRun.Status.Steps[i].Terminated.Message` of type `PipelineResourceResult`. Whenever we can find the two error results in a step's message field, we surface the error as `failureDetails`.
+
+So far, we have presented a mechanism to lift errors results into a BuildRun. However, no component has produced error results yet.  One source of failure is misconfiguration in the source step for git applications. For example, a user can forget to provide authentication details for a private repository. Therefore, we propose to implement error results for `cmd/git`. On top of that, we propose implementing smarter error reporting by parsing and classifying git's messages for errors in `stdout`. A third-party application would be able to provide mitigations for failed builds based on error classes in git, for example, a UI workflow to reenter basic authentication details.
+
+The following errors classes are implemented for git:
+- `AuthInvalidUserOrPass`, expresses that basic authentication is not possible
+- `AuthInvalidKey`, expresses that ssh authentication is not possible
+- `BranchNotFound` expresses that a remote branch does not exist
+- `RepositoryNotFound`, expresses that the remote target for the git operation does not exist. It triggers when an error message is enough to determine that the remote target does not exist and is mostly derived from the Git server's messages e.g. GitLab or GitHub
+- `AuthPrompted`, is caused when a repo is not found, is private and authentication is insufficient
+- `Unknown`, is the class of choice if no other class fits
+
+
+Each class has a helpful, static message to hint a user to a right direction to fix his build. The unknown class is generic and chosen if no other class matches. Its message is an exception since we can not determine the exact class. Thus, we propagate the git's stdout and publish it as a message.
```

Comment:
> Thus, we propagate the git's stdout and publish it as a message.

This might be too long for a result. Maybe we need to limit the maximum length of what we pass here, or, as suggested above as an alternative approach, we simply do not set the results. Then the user is guided to look at the logs and that's it.

Reply:
Addressed in: https://github.com/shipwright-io/build/pull/901/files#diff-7fc17d01c94992ca783a5e0345dd88a503c0c9b48e2ff05eabf6bc14821f2825R463. The error message can at max take about a quarter of the allowed total size as of now.